Custom URL Scheme Implementation

hassan31 opened the pull request:
Implemented custom url scheme to support launching the Quran Application from outside the application
For example, enter this in browser `quran-ios://` and it will redirect you to the Quran Application

Example Video

https://github.com/user-attachments/assets/cafcc9f4-5c24-4ebd-b6fb-1b680ab69f7c

iman2153 reviewed (approved):
lgtm

hassan31 commented:
@mohamede1945 could you please review this pr and approve so that it can be merged. Thanks

mohamede1945 commented:
Salam alikom Muhammad, thank you for the contribution. But the code is empty. There's nothing implemented to parse the query parameters and navigate to a specific Ayah as mentioned in #348.

hassan31 commented:
@mohamede1945 Wa Alaikumuslaam, Yes, this is the partial implementation, in which I have launched the application from outside link. In followup pr I will write the code to parse. So if you could approve and merge that would be really great.

mohamede1945 commented on Example/QuranEngineApp/Classes/AppDelegate.swift:
```diff
@@ -49,6 +49,27 @@ class AppDelegate: UIResponder, UIApplicationDelegate {
+    
+    func application(_ app: UIApplication, open url: URL, options: [UIApplication.OpenURLOptionsKey: Any] = [:]) -> Bool {
+        // Check if the URL scheme matches the custom scheme
+        if url.scheme == "quran-ios" {
```
Let's handle "quran" and "quran-ios" schemes please.

hassan31 replied:
addressed

mohamede1945 commented on Example/QuranEngineApp/Classes/AppDelegate.swift:
```diff
@@ -49,6 +49,27 @@ class AppDelegate: UIResponder, UIApplicationDelegate {
     }
+    
+    func application(_ app: UIApplication, open url: URL, options: [UIApplication.OpenURLOptionsKey: Any] = [:]) -> Bool {
+        // Check if the URL scheme matches the custom scheme
```
Please remove this redundant comment. The code is clear. Same with the other inline comments below.

hassan31 replied:
addressed

mohamede1945 commented on Example/QuranEngineApp/Classes/AppDelegate.swift:
```diff
@@ -49,6 +49,27 @@ class AppDelegate: UIResponder, UIApplicationDelegate {
+        // Check if the URL scheme matches the custom scheme
+        if url.scheme == "quran-ios" {
+            // Parse the URL to determine the desired action
+            let path = url.host ?? ""
```
host is not the path. We should look at the url.path rather than the host.

hassan31 replied:
addressed

mohamede1945 commented on Example/QuranEngineApp/Classes/AppDelegate.swift:
```diff
@@ -49,6 +49,27 @@ class AppDelegate: UIResponder, UIApplicationDelegate {
+        return false
+    }
+
+    private func handleCustomURL(path: String) {
```
`handle` seems very generic. I would rename it to `navigateTo(path: String)`. Also let's make this method return false if it can't parse and navigate to the path.

hassan31 replied:
addressed

mohamede1945 commented on Example/QuranEngineApp/Info.plist:
```diff
@@ -2,6 +2,17 @@
+			<string>com.quran.QuranEngineApp</string>
+			<key>CFBundleURLSchemes</key>
+			<array>
+				<string>quran-ios</string>
```
Let's add an entry for `quran` scheme please.

hassan31 replied:
addressed

mohamede1945 reviewed (changes requested):
Sounds good. I've added a couple of comments, please take a look. Also, please make sure to run `make format-autocorrect` to fix formatting issues.

Thank you!

mohamede1945 commented:
Oh actually, please move the validation, parsing and navigation code to the QuranEngine instead of the example app. The example app would simply just implement the delegate method but then calls the QuranEngine to do the actual logic, this way the code can be shared among different apps.

hassan31 commented:
> Oh actually, please move the validation, parsing and navigation code to the QuranEngine instead of the example app. The example app would simply just implement the delegate method but then calls the QuranEngine to do the actual logic, this way the code can be shared among different apps.

AOA @mohamede1945 I have addressed your comments, kindly review again. Also regarding move the validation, parsing and navigation code to QuranEngine, i can see the same code there, but in pr it reflects in Example/QuranEngine app. May be if you guide how to move then I would do. Please check below screenshot.

<img width="1296" alt="Screenshot 2024-09-08 at 12 16 07 AM" src="https://github.com/user-attachments/assets/6bfa531b-0ea5-4635-8bae-1c864ffc194d">

mohamede1945 commented:
You see [here](https://github.com/quran/quran-ios/blob/main/Example/QuranEngineApp/Classes/AppDelegate.swift#L26-L29) how the code in the example project delegates the business logic to the DownloadManager and ReadingResourcesService. These live in the [QuranEngine library](https://github.com/quran/quran-ios/blob/main/Package.swift) not in the Example project. The example project just delegates the call to these services.

In url scheme navigation logic, I think it makes sense to live in the AppStructureFeature.

hassan31 commented:
> You see [here](https://github.com/quran/quran-ios/blob/main/Example/QuranEngineApp/Classes/AppDelegate.swift#L26-L29) how the code in the example project delegates the business logic to the DownloadManager and ReadingResourcesService. These live in the [QuranEngine library](https://github.com/quran/quran-ios/blob/main/Package.swift) not in the Example project. The example project just delegates the call to these services.
> 
> In url scheme navigation logic, I think it makes sense to live in the AppStructureFeature.

@mohamede1945 please review now, is this what you are expecting to move the logic to QuranEngine?

hassan31 commented:
> You see [here](https://github.com/quran/quran-ios/blob/main/Example/QuranEngineApp/Classes/AppDelegate.swift#L26-L29) how the code in the example project delegates the business logic to the DownloadManager and ReadingResourcesService. These live in the [QuranEngine library](https://github.com/quran/quran-ios/blob/main/Package.swift) not in the Example project. The example project just delegates the call to these services.
> 
> In url scheme navigation logic, I think it makes sense to live in the AppStructureFeature.

@mohamede1945 can we connect through some communication tools, like slack or google meet, I wanted to learn more about the structure?

hassan31 commented:
@mohamede1945 are you still available to review the latest changes?

mohamede1945 commented:
Jazak Allah khyrn for the changes. I'm going to merge it after CI finishes. You can reach out to me on discord with the same username used here in github.

hassan31 commented:
> Jazak Allah khyrn for the changes. I'm going to merge it after CI finishes. You can reach out to me on discord with the same username used here in github.

Merge is blocked, because of some checks are not successful. I also installed the swiftlint, but looks good on my side. May be you can help to find what is wrong with which format.

Also, I sent you request on discord. Please accept. Jazakallah Khair.
